### DISCUSSION TITLE
complete noob - form_with not producing forms

### DISCUSSION THREAD

**Discussion by thisIsLoading** [yippee-fun/phlex#818] [Help]
Hi, this is my very first attempt at phlex in rails (phlex 2). and my small brain doesnt help

i struggle completely with the simplest tasks. in this case i try to create a form:

```ruby
class Components::Table::HeaderForm < Components::Base
  include Phlex::Rails::Helpers::FormWith

  def initialize(filter:, params:, path:, buttons:)
    @filter  = filter
    @params  = params
    @path    = path
    @buttons = buttons
  end

  def view_template
    form_with(url:   @path, method: :get, id: "opportunities-filter-form",
              class: "flex items-center space-x-4", data: {
        turbo_stream: true,
        controller:   "form advanced-filter",
        turbo_action: "advance",
        action:       "input->form#submit",
        update_url:   true
      }) do |form|

      div(class: "flex items-center space-x-2") do
        render Components::Shared::TimeRangeButtons.new(
          form:      form,
          attribute: :time_range,
          options:   [ [ "1d", "1 D" ], [ "7d", "7 D" ], [ "30d", "30 D" ], [ "90d", "90 D" ] ]
        )
        form.hidden_field(name: "sort", value: @params[:sort], data: { sort_link_target: "sort" })
        form.hidden_field(name: "direction", value: @params[:direction], data: { sort_link_target: "direction" })
        form.hidden_field(name: "limit", value: @params[:limit])
      end
    end
  end
end
```

which kinda renders as expected but under the hood has no form object in the dom. it just renders as if it were all outside a form?

what do i do wrong?

bonus question:

my poor attempt at the the time range buttons component:

```ruby
class Components::Shared::TimeRangeButtons < Components::Base
  def initialize(form:, attribute:, options:)
    @form      = form
    @attribute = attribute
    @options   = options
  end

  def view_template
    div(class: "isolate inline-flex rounded shadow-sm w-64") do
      @options.each_with_index do |(value, label), index|
        button_class = case index
                         when 0 then "rounded-l"
                         when @options.size - 1 then "rounded-r"
                         else ""
                       end

        # is_active = @form.object.send(@attribute) == value
        is_active = false

        # Render the radio button input
        @form.radio_button(@attribute, value, id: "#{value}_radio", class: "hidden", data: { action: "change->form#submit" })

        # Render the label for the button
        label(for: "#{value}_radio", class: "cursor-pointer text-center inline-flex justify-center items-center btn-group flex-1 #{'btn-primary' if is_active} #{button_class}") do
          label
        end
      end
    end
  end
end
```

fails as well as `@form.object` is nil, so the is_active check is failing with an undefined method for nil. it worked back when it was a rails partial, so i guess i have to do that differently now, but its very hard to find infos regarding this. i googled my ass off.

any help is very much appreciated!

**Comment by antonioJASR** [comment]
I would try using `helpers.form_with` instead.

**Comment by joeldrapper** [comment]
It’s possible our adapter for `form.object` is wrong. In Phlex, methods like `form.hidden_field` need to push the return value to the DOM and return `nil`. But `form.object` should return the return value.

**Reply by joeldrapper** [reply]
No, it looks like we do have `form.object` mapped to just return its value. https://github.com/phlex-ruby/phlex-rails/blob/80388e7c1a5ba53f11e8de45eeb250bb99bc9060/lib/phlex/rails/buffered_form_builder.rb#L32

Can you make a minimal reproduction in a repo I can look at?

**Reply by joeldrapper** [reply]
Obviously if there’s a bug here, we want to fix it, but aside from that, you’ll probably have a much better time building forms with components and regular HTML than using the form builder. The form builder doesn’t really get you any advantage if you already have Phlex.
